packages/next/src/client/app-index.tsx

```diff
@@ -154,60 +153,14 @@ function Root({ children }: React.PropsWithChildren<{}>): React.ReactElement {
 }
 
 export function hydrate() {
-  if (process.env.NODE_ENV !== 'production') {
-    const rootLayoutMissingTagsError = (self as any)
-      .__next_root_layout_missing_tags_error
-    const HotReload: typeof import('./components/react-dev-overlay/app/hot-reloader-client').default =
-      require('./components/react-dev-overlay/app/hot-reloader-client')
-        .default as typeof import('./components/react-dev-overlay/app/hot-reloader-client').default
-
-    // Don't try to hydrate if root layout is missing required tags, render error instead
-    if (rootLayoutMissingTagsError) {
-      const reactRootElement = document.createElement('div')
-      document.body.appendChild(reactRootElement)
-      const reactRoot = (ReactDOMClient as any).createRoot(reactRootElement, {
-        onRecoverableError,
-      })
-
-      reactRoot.render(
-        <GlobalLayoutRouterContext.Provider
-          value={{
-            buildId: 'development',
-            tree: rootLayoutMissingTagsError.tree,
-            changeByServerResponse: () => {},
-            focusAndScrollRef: {
-              apply: false,
-              onlyHashChange: false,
-              hashFragment: null,
-              segmentPaths: [],
-            },
-            nextUrl: null,
-          }}
-        >
-          <HotReload
-            assetPrefix={rootLayoutMissingTagsError.assetPrefix}
-            // initialState={{
-            //   rootLayoutMissingTagsError: {
-            //     missingTags: rootLayoutMissingTagsError.missingTags,
-            //   },
-            // }}
-          />
-        </GlobalLayoutRouterContext.Provider>
-      )
-
-      return
-    }
-  }
-
   const actionQueue = createMutableActionQueue()
 
+  const rootLayoutMissingTags = (self as any)
+    .__next_root_layout_missing_tags as ('html' | 'body')[] | undefined
+
```

Comment:
Rather than using `as` here, creating a guard function would be safer ([see this TS Playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgLLAM4dAcwBIRwAmA8lAEID2RAnsgN4BQyLyA+myBAB5htSVKfADZwalAK58AtpmwgcbMHBwYAXMgBKhIpRDCaAQShQxAHgBEACx0WAPhYBG1GhYB8jAL6NGMCSAQwYD1kBBsEAGsACgwIYRgNOBAaAEoNWPjkTDQ5XAJiMipaBmZWKAgwCSgQZAByDi5efkERMUkZXIUlFQxarJqMmGQAMmHkY1MaADpMCbEYuJgphp4+ASE2UXEpNlksXG7VFK8fBD0MMGQMZABeBm9GYCGosIhImJSUktZkAHpf5AAOUoAHdkCCUAgklc4DA4nREEgsMgwDZkAAHATo6BgaalFgYZacVbNDZbdq7TqKZSqE6Mf7IAACYAwAFoeNjAuyTJQoIxCSsmutWtsOvsujSMEA))

Reply:
Or an alternative is to augment the `Window` type [`7af4a43` (#62815)](https://github.com/vercel/next.js/pull/62815/commits/7af4a43ab25d98b386c7cea09efc8a1e0371732b)